### PR TITLE
[masic support] 'show run bgp' support for multi-asic

### PR DESCRIPTION
**Pull request by wenyiz2021**
Support 'show run bgp' on multi-asic.
if 'show run bgp' is given, it'll print all running bgp configs from all asics
if a namespace is given, will print only for that asic.

#### What I did

1. take option 'namespace' for 'show run bgp' cmd:
  ```
admin@:~$ show run bgp -h
Usage: show run bgp [OPTIONS]

  Show BGP running configuration

Options:
  --verbose             Enable verbose output
  -n, --namespace TEXT  Option needed for multi-asic only: provide namespace
                        name
  -?, -h, --help        Show this message and exit.
```
2. if namespace not given, show run bgp on all asics, 1 by 1
3.  for each asic, there is title indicates following info is for which asic
4. single-asic behavior remains the same from 'show run bgp'

#### How to verify it
verified on a multi-asic device, sample output:
```
admin@str-masic-acs-1:~$ show run bgp

------------Showing running config bgp on asic0------------
Building configuration...

Current configuration:
.....

------------Showing running config bgp on asic1------------
Building configuration...

Current configuration:
!
frr version 7.2.1-sonic
frr defaults traditional
....
```

```
admin@svcstr-masic-acs-1:~$ show run bgp -n asic3

------------Showing running config bgp on asic3------------
Building configuration...

Current configuration:
!
frr version 7.2.1-sonic
frr defaults traditional
.....
!
line vty
!
end
```

single-asic o/p is not changed:
After:
```
admin@str2-7050cx3-acs-01:/usr$ show run bgp
Building configuration...

Current configuration:
!
frr version 7.5.1-sonic
frr defaults traditional
hostname str2-7050cx3-acs-01
log syslog informational
log facility local4
agentx
no service integrated-vtysh-config
....
```

```
admin@str2-sasic-acs-02:/usr$ show run bgp -n asic0
Usage: show run bgp [OPTIONS]
Try "show run bgp -h" for help.

Error: -n/--namespace is not available for single asic
```

#### Previous command output (if the output of a command-line utility has changed)
```
admin@svcstr-masic-acs-1:~$ show run bgp -n asic3
Error: no such option: -n
admin@svcstr-masic-acs-1:~$ show run bgp
Usage: /usr/bin/vtysh -n [0 to 5] [OPTION]... 
admin@svcstr-masic-acs-1:~$ show run bgp -n 3
Error: no such option: -n
```

### CODE REVIEW

File: utilities_common/multi_asic.py [hunk 1]
```diff
@@ -112,6 +112,17 @@ def multi_asic_display_default_option():
 ]
 
 def multi_asic_namespace_validation_callback(ctx, param, value):
+    # TODO: this condition has a bug: it should be either "multi_asic.is_multi_asic()"
+    # or "self.is_multi_asic", but changing it to correct version will lead to the 
+    # following cli not working on single-asic:
+    # show ip(v6) bgp neighbors/network, show starm_control add/del
```
Comment (gechiang) [inline utilities_common/multi_asic.py:118]: What is "show starm_control add/del"? is this existing CLI?

Comment (wenyiz2021) [reply utilities_common/multi_asic.py:118]: sorry for typo, should be "show storm_control"

File: show/main.py [hunk 2]
```diff
@@ -1385,10 +1385,37 @@ def ports(portname, verbose):
...
-def bgp(verbose):
+@click.option('--namespace', '-n', 'namespace', required=False, default=None, type=str, show_default=False, 
+              help='Option needed for multi-asic only - provide namespace name or all')
+def bgp(namespace, verbose):
     """Show BGP running configuration"""
-    cmd = 'sudo {} -c "show running-config"'.format(constants.RVTYSH_COMMAND)
-    run_command(cmd, display_cmd=verbose)
+    if multi_asic.is_multi_asic() and (namespace not in multi_asic.get_namespace_list() or namespace != 'all'):
+        ctx = click.get_current_context()
+        ctx.fail("-n/--namespace option required. provide namespace from list {}, or give '-n all'".format(multi_asic.get_namespace_list()))
```
Comment (gechiang) [inline show/main.py:1394]: Would it be better to have default for multi-asic whereif the user do not specify anything such as "-n all" it defaults to display for all ASICs instead of forcing the user to have to specify "-n all"?
What does the other multi-asic show commands behave when it comes to BGP related show commands? Do they all forces the user to always have to specify -n option?  We should keep similar default behavior...

Comment (wenyiz2021) [reply show/main.py:1394]: @gechiang your point sounds fair. Other cmds can run without 'namespace' specified and show all.
But here is the blocker:
- other cmd like `show ip interfaces` uses `multi_asic_click_options` that can take optional namespaces, while this also gives a 'display' option that rvtysh cmd cannot take for `show run bgp`
- even with `required = false` inside option `-n`, it won't work as an optional option, it still requires '-n' to be given for multi-asic as a must have.

will try to see how to make `-n` optional.

Comment (wenyiz2021) [reply show/main.py]: This issue is addressed.
`show run bgp` by default will print for all asics.
`show run bgp -n asicx` will print for asicx

**Comment by wenyiz2021** [comment]
@gechiang @arlakshm can you please help to review this PR? thanks

**Review by gechiang — DISMISSED** [review]
LGTM

File: show/main.py [hunk 3]
```diff
@@ -1435,10 +1435,40 @@ def ports(portname, verbose):
...
+
+    output = ""
+    cmd = "show running-config bgp"
+    import utilities_common.bgp_util as bgp_util
+    if multi_asic.is_multi_asic():
+        if not namespace:
+            ns_list = multi_asic.get_namespace_list()
+            for ns in ns_list:
+                output += "\n------------Showing running config bgp on {}------------\n".format(ns)
+                output += bgp_util.run_bgp_command(cmd, ns)
```
Comment (arlakshm) [inline show/main.py:1465]: can we use `run_bgp_show_command` iso `run_bgp_command` ? `run_bgp_show_command` has the correct permission to do show commands for users with RO privilages

File: tests/show_run_bgp_test.py [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+        assert result.exit_code == 0
+        assert result.output == show_run_bgp_masic_asic0
+
+    @classmethod
+    def teardown_class(cls):
+        print("TEARDOWN")
+        from .mock_tables import mock_single_asic
+        importlib.reload(mock_single_asic)
+        from .mock_tables import dbconnector
+        dbconnector.load_database_config
```
Comment (arlakshm) [inline tests/show_run_bgp_test.py:209]: can you add some tests for negative tests like bgp container is not running.

File: utilities_common/multi_asic.py [hunk 5]
```diff
@@ -114,6 +114,17 @@ def multi_asic_display_default_option():
...
+    # TODO: this condition has a bug: it should be either "multi_asic.is_multi_asic()"
+    # or "self.is_multi_asic", but changing it to correct version will lead to the 
+    # following cli not working on single-asic:
+    # show ip(v6) bgp neighbors/network, show storm_control add/del
+    # e.g. 
+    # admin@str2-7050cx3-acs-02:/usr$ show ip bgp neighbor -n asic0
+    # -n/--namespace is not available for single asic
+    # Aborted!
+    # admin@str2-7050cx3-acs-02:/usr$ show ip bgp neighbor
+    # -n/--namespace is not available for single asic
```
Comment (arlakshm) [inline utilities_common/multi_asic.py:126]: are you fixing this in another PR? Can you open a github issue iso adding comment in the code

**Review by arlakshm — COMMENTED** [review]
as comments

File: utilities_common/multi_asic.py (hunk 5 above)
Comment (wenyiz2021) [reply utilities_common/multi_asic.py]: raised issue: https://github.com/sonic-net/sonic-buildimage/issues/13560

File: show/main.py (hunk 3 above)
Comment (wenyiz2021) [reply show/main.py]: done

File: tests/show_run_bgp_test.py (hunk 4 above)
Comment (wenyiz2021) [reply tests/show_run_bgp_test.py:228]: Hi @arlakshm , added neg UT

**Review by arlakshm — APPROVED** [review]
